internal/defence/limiter_test.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package defence
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	"storj.io/storj/internal/testcontext"
+)
+
+const (
+	key1           = "email1@example.com"
+	key2           = "email2@example.com"
+	maxAttempts    = 3
+	attemptsPeriod = time.Second
+	banDuration    = time.Second
+	clearPeriod    = time.Second * 3
+)
+
+func TestLimiter(t *testing.T) {
+	var limiter = NewLimiter(maxAttempts, attemptsPeriod, banDuration, clearPeriod)
+
+	t.Run("Testing constructor", func(t *testing.T) {
+		assert.Equal(t, limiter.attempts, maxAttempts)
+		assert.Equal(t, limiter.attemptsPeriod, attemptsPeriod)
+		assert.Equal(t, limiter.lockDuration, banDuration)
+		assert.NotNil(t, limiter.attempts)
+	})
+
+	t.Run("should not be banned when to attack < attempts during attemptsPeriod", func(t *testing.T) {
+		result := false
+
+		for i := 0; i < 2; i++ {
+			result = limiter.Limit(key1)
+		}
+
+		assert.Equal(t, result, true)
+	})
+
+	t.Run("should be banned when attack > attempts when attemptsPeriod exceeded", func(t *testing.T) {
+		result := false
+
+		for i := 0; i <= maxAttempts; i++ {
+			result = limiter.Limit(key2)
+		}
+
+		assert.Equal(t, result, false)
+	})
+
+	limiter.Close()
+}
+
+func TestLimiterConcurrent(t *testing.T) {
+	ctx, cancel := context.WithCancel(testcontext.New(t))
```

Comment:
I think you need a `defer ctx.Cleanup()` on the `ctx` value returned by `testcontext.New`.